contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,850 @@
+            _pool = _validators[operatorAddresses[i]].creditContract;
+            require(_pool != address(0), "validator not exist");
+            validatorPools[i] = _pool;
+        }
+
+        IGovBNB(GOV_TOKEN_ADDR).sync(validatorPools, account);
+    }
+
+    /*----------------- system functions -----------------*/
+    function distributeReward(address consensusAddress) external payable onlyValidatorContract {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        Validator memory valInfo = _validators[operatorAddress];
+        if (valInfo.creditContract == address(0) || valInfo.jailed) {
+            emit RewardDistributeFailed(operatorAddress, "INVALID_VALIDATOR");
+            return;
+        }
+
+        try IStakeCredit(valInfo.creditContract).distributeReward{ value: msg.value }(valInfo.commission.rate) {
+            emit RewardDistributed(operatorAddress, msg.value);
+        } catch Error(string memory reason) {
+            emit RewardDistributeFailed(operatorAddress, bytes(reason));
+        } catch (bytes memory lowLevelData) {
+            emit RewardDistributeFailed(operatorAddress, lowLevelData);
+        }
+    }
+
+    /**
+     * @dev Get new eligible validators from consensus engine
+     */
+    function updateEligibleValidatorSet(
+        address[] calldata validators,
+        uint64[] calldata votingPowers,
+        bytes[] calldata voteAddrs
+    ) external onlyCoinbase onlyZeroGasPrice {
+        uint256 newLength = validators.length;
+        if (newLength == 0) {
+            return;
+        }
+        uint256 oldLength = _eligibleValidatorSet.length;
+        if (oldLength > newLength) {
+            for (uint256 i = newLength; i < oldLength; ++i) {
+                _eligibleValidatorSet.pop();
+            }
+        }
+
+        uint256 j;
+        IBSCValidatorSet.Validator[] memory bscValidatorSet = new IBSCValidatorSet.Validator[](newLength);
+        for (uint256 i; i < newLength; ++i) {
+            if (j >= oldLength) {
+                _eligibleValidatorSet.push(validators[i]);
+            } else {
+                _eligibleValidatorSet[j] = validators[i];
+            }
+            ++j;
+
+            bscValidatorSet[i] = IBSCValidatorSet.Validator({
+                consensusAddress: validators[i],
+                feeAddress: payable(address(0)),
+                BBCFeeAddress: address(0),
+                votingPower: votingPowers[i],
+                jailed: false,
+                incoming: 0
+            });
+        }
+
+        // update validator set in BSCValidatorSet contract
+        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).updateValidatorSetV2(bscValidatorSet, voteAddrs);
+    }
+
+    function downtimeSlash(address consensusAddress) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
```

Comment:
Please don't revert, because this is invoked by consensus engine

Reply:
It's warped in `SlashIndicator`
```
  function _downtimeSlash(address validator, uint256 count) internal {
    try IStakeHub(STAKE_HUB_ADDR).downtimeSlash(validator) {
    } catch Error(string memory reason) {
      emit failedFelony(validator, count, bytes(reason));
    } catch (bytes memory lowLevelData) {
      emit failedFelony(validator, count, lowLevelData);
    }
  }
```